### PR TITLE
clang/OpenCL: set sqrt fp accuracy on call to Z4sqrt

### PR DESCRIPTION
**Pull request by rjodinchr**
This is reverting the previous implementation to avoid adding inline function in opencl headers.
This was breaking clspv flow google/clspv#1231, while https://reviews.llvm.org/D156743 mentioned that just decorating the call node with `!pfmath` was enough.
This PR is implementing this idea.
The test has been updated with this implementation.

### CODE REVIEW

**Comment by rjodinchr** [comment]
@alan-baker @arsenm @AnastasiaStulova 
Could you review this PR please?
Thank you

File: clang/lib/CodeGen/CGCall.cpp [hunk 1]
```diff
@@ -5612,6 +5612,10 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
                               BundleList);
     EmitBlock(Cont);
   }
+  if (CI->getCalledFunction() && CI->getCalledFunction()->hasName() &&
+      CI->getCalledFunction()->getName().contains("Z4sqrt")) {
```
Comment (arsenm) [inline clang/lib/CodeGen/CGCall.cpp:5616]: Probably should check the language, and exact match on the permissible names

Comment (rjodinchr) [reply clang/lib/CodeGen/CGCall.cpp:5616]: the language is checked by `SetSqrtFPAccuracy`. I'll have a look at a better match for the name

**Comment by rjodinchr** [comment]
@arsenm Could you review this PR again please?

File: clang/lib/CodeGen/CGCall.cpp [hunk 2]
```diff
@@ -5612,6 +5612,10 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
                               BundleList);
     EmitBlock(Cont);
   }
+  if (CI->getCalledFunction() && CI->getCalledFunction()->hasName() &&
+      CI->getCalledFunction()->getName().startswith("_Z4sqrt")) {
```
Comment (arsenm) [inline clang/lib/CodeGen/CGCall.cpp:5616]: This isn't a specific enough filter. Exactly match the full name for all the types?

Comment (arsenm) [reply clang/lib/CodeGen/CGCall.cpp:5616]: Maybe it's good enough, I guess the Z4 filters out other prefixes. Not sure about arbitrary custom ext_vector_type

Comment (rjodinchr) [reply clang/lib/CodeGen/CGCall.cpp:5616]: Using a custom `ext_vector_type` in OpenCL feels like an undefined behaviour. I don't see anything about it in the specification.
Am I missing something?

Comment (rjodinchr) [reply clang/lib/CodeGen/CGCall.cpp:5616]: @arsenm ?

**Comment by rjodinchr** [comment]
@arsenm could you have another look at this PR?
Thank you

**Comment by alan-baker** [comment]
This is preferable to defining the function in the OpenCL headers (as was noted in the original PR). To me, the important part is covering the OpenCL use case. So if we want to only check the overloads OpenCL generates I think that would be ok, but I expect this generic implementation is sufficient.

File: clang/lib/CodeGen/CGCall.cpp (hunk 2 above)
Comment (arsenm) [reply clang/lib/CodeGen/CGCall.cpp:5616]: ext_vector_type is an implementation detail. The standardized vector types just happen to be typedefs of ext_vector_type.  It's not undefined behavior, it's just something out of spec you can do

**Review by arsenm — APPROVED** [review]
lgtm, but a strict reading of the spec would filter out arbitrary other ext_vector_types

**Comment by rjodinchr** [comment]
What is the next step to get this PR merged?

**Comment by rjodinchr** [comment]
@AnastasiaStulova Could you please take a look? thanks

**Pull request merged**